README.md
# OctoCostToo

[![hacs_badge](https://img.shields.io/badge/HACS-Custom-orange.svg)](https://github.com/custom-components/hacs) [![codecov](https://codecov.io/gh/lildude/OctoCostToo/branch/main/graph/badge.svg?token=OEGCXZNPDX)](https://codecov.io/gh/lildude/OctoCostToo)

Please consider signing up to Octopus Energy with my referral code: <https://share.octopus.energy/mauve-ash-375> to give you and me £50 credit in the mean time.

Want to motivate me to improve this quicker? [Sponsor me](https://github.com/sponsors/lildude) to work on it 😉 💖

---

## Summary

OctoCostToo is an [AppDaemon](https://www.home-assistant.io/docs/ecosystem/appdaemon/) app for [Home Assistant](https://www.home-assistant.io/) which calculates the daily, monthly and yearly cost and usage of the Octopus Energy tariffs. This is a refactored and improved fork of the now archived [octocost](https://github.com/badguy99/octocost).

By default OctoCostToo only gathers usage and cost information for the current (as of Jan 2021) Agile tariff: `AGILE-18-02-21` but can be configured to gather usage and cost information for an another electricity tariff, giving you the opportunity to compare your current tariff to the Agile tariff.

Usage and cost information can also be gathered for a gas tariff.

OctoCostToo creates and sets sensors for daily, monthly and yearly cost (£) and usage (kWh), up to and including yesterday:

```yaml
sensor.octopus_daily_cost
sensor.octopus_daily_usage
sensor.octopus_monthly_cost
sensor.octopus_monthly_usage
sensor.octopus_yearly_cost
sensor.octopus_yearly_usage
```

If a comparison electricity tariff is configured, OctoCostToo will also create the following sensors:

```yaml
sensor.octopus_comparison_daily_cost
sensor.octopus_comparison_monthly_cost
sensor.octopus_comparison_yearly_cost
```

If the gas section is configured, OctoCostToo will also create the following sensors:

```yaml
sensor.octopus_monthly_gas_cost
sensor.octopus_monthly_gas_usage
sensor.octopus_yearly_gas_cost
sensor.octopus_yearly_gas_usage
```

The data is updated once every two hours, although in reality the data Octopus Energy gets only seems to be updated once a day, so this is a compromise between trying to be up-to-date, and not hammering their servers, when the data doesn't update very frequently anyway.

## Installation

Use [HACS](https://github.com/custom-components/hacs) or [download the tarball](https://github.com/lildude/OctoCostToo/releases) and extract the `OctoCostToo` directory from inside the `apps` directory to your local `apps` directory, then add the configuration to enable the OctoCostToo module.

## Apps.yaml Configuration

```yaml
OctoCostToo:
  module: octocosttoo 
  class: OctoCostToo 
  region: H
  mpan: <13 digit MPAN number>
  serial:  <Serial number>
  auth: <Octopus Energy API Key>
  start_date: 2020-02-23
  comparison_tariff: FIX-12M-20-02-12
  gas:
    mprn: <Gas MPRN number>
    gas_serial: <Gas meter serial number>
    gas_tariff: FIX-12M-20-02-12
    gas_start_date: 2020-02-23
```

The module and class sections need to remain as above, other sections should be changed as required. The whole gas section is optional and can be excluded if not required.

| Field             | Changeable | Example          |
| -----             | ---------- | -------          |
| Title             | Yes        | octocosttoo      |
| module            | No         | octocosttoo      |
| class             | No         | OctoCostToo      |
| region            | Yes        | H                |
| mpan              | Yes        | 2000012345678    |
| serial            | Yes        | 20L123456        |
| auth              | Yes        | sk_live_abcdefg  |
| start_date        | Yes        | 2020-02-23       |
| comparison_tariff | Yes        | FIX-12M-20-02-12 |
| gas:              | Yes        |                  |
| mprn              | Yes        | 1234567890       |
| gas_serial        | Yes        | E1S12345678901   |
| gas_tariff        | Yes        | FIX-12M-20-02-12 |
| gas_start_date    | Yes        | 2020-02-23       |

The `start_date` setting should be set to the date you started on the Agile Octopus tariff, not the date you joined Octopus Energy. It is used to adjust the start point if you joined within the current year or month, it should not be left blank if you joined earlier.
`region` is the region letter from the end of `E-1R-AGILE-18-02-21-H` which can be found on the [Octopus Energy developer dashboard](https://octopus.energy/dashboard/developer/) webpage in the Unit Rates section for your account.

## Limitations

- OctoCostToo only caters for single-rate comparison electricity and gas tariffs.

### Lovelace UI Cards

Once the sensors are created, they can be displayed as cards within the Lovelace UI. For example:

```yaml
- entities:
    - entity: sensor.octopus_yearly_usage
      icon: 'mdi:flash'
      name: Yearly Usage (kWh)
    - entity: sensor.octopus_yearly_cost
      icon: 'mdi:cash'
      name: Yearly Cost (£)
    - entity: sensor.octopus_monthly_usage
      icon: 'mdi:flash'
      name: Monthly Usage (kWh)
    - entity: sensor.octopus_monthly_cost
      icon: 'mdi:cash'
      name: Monthly Cost (£)
  show_icon: true
  title: Octopus Usage / Cost
  type: glance
```

![Example Lovelace UI Usage and Cost glance card](https://github.com/lildude/OctoCostToo/blob/main/LovelaceUsageCard.PNG)

## Contributing

Want to contribute to this project? Great! Fork the repo, make your changes (don't forget to add tests 😉) and submit a pull request.

You can install the necessary dependencies and run the tests locally as follows:

```console
$ python -m pip install --upgrade pip
$ pip install -r requirements-test.txt
$ pytest
```

## Credit

OctoCostToo is a fork of the original [octocost](https://github.com/badguy99/octocost).

**NOTE:** The configuration for OctoCostToo is not compatible with the original.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
Changes are only recorded from 0.2.0 when the project was forked from the archived upstream project.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2021-03-08

### Added

- Support for a fixed-rate tariff for easy comparison with the Agile tariff.
- Unit tests.

### Changed

- Renamed the project to OctoCostToo after forking.
- Many of the original configuration options have been renamed to use snake_case.
- Gas usage used to determine the cost in £ per kWh is converted from m3 to kWh using the formula: `X × 1.02264 × 39.0 ÷ 3.6` or simply `X × 11.0786`.
- Refactored a lot of the code to reduce duplication and improve maintenance.

### Fixed

- Standing charge for gas costing is now included in the cost calculations.

[Unreleased]: https://github.com/lildude/OctoCostToo/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/lildude/OctoCostToo/compare/v0.1.9...v0.2.0


tests/conftest.py
import pytest
import requests


# Prevent accidental outgoing network requests - we should be mocking these
@pytest.fixture(autouse=True)
def no_http_requests(monkeypatch):
    """
    Prevent accidental HTTP requests during testing
    """

    def urlopen_mock(self, method, url, *args, **kwargs):
        raise RuntimeError(
            f"The test was about to {method} {self.scheme}://{self.host}{url}"
        )

    monkeypatch.setattr(
        "urllib3.connectionpool.HTTPConnectionPool.urlopen", urlopen_mock
    )


@pytest.fixture(name="mock_elec_meter_points")
def fixture_mock_elec_meter_points(requests_mock):
    """
    Mock response for getting meter points used to determine the region
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/electricity-meter-points/12345",
        text=open("tests/fixtures/elec-meter-points.json", "r").read(),
    )


@pytest.fixture(name="mock_elec_consumption_one_day")
def fixture_mock_elec_consumption_one_day(requests_mock):
    """
    Mock response for a single day's electricity usage
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/electricity-meter-points/12345/meters/67890/consumption/?order_by=period&period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z&page_size=47",
        text=open("tests/fixtures/elec-one-day-usage.json", "r").read(),
    )


@pytest.fixture(name="mock_elec_agile_cost_one_day")
def fixture_mock_elec_agile_cost_one_day(requests_mock):
    """
    Mock response for a single day's Agile electricity prices
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/products/AGILE-18-02-21/electricity-tariffs/E-1R-AGILE-18-02-21-H/standard-unit-rates/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/agile-cost-one-day.json", "r").read(),
    )


@pytest.fixture(name="mock_elec_fixed_rates")
def fixture_mock_elec_fixed_rates(requests_mock):
    """
    Mock response for electricity standard unit rates & standing charge
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/electricity-tariffs/E-1R-FIX-12M-20-09-21-H/standard-unit-rates/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/elec-standard-rate.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/electricity-tariffs/E-1R-FIX-12M-20-09-21-H/standing-charges/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/elec-standing-charges.json", "r").read(),
    )


@pytest.fixture(name="mock_gas_consumption_one_day")
def fixture_mock_gas_consumption_one_day(requests_mock):
    """
    Mock response for a single day's gas usage
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/gas-meter-points/54321/meters/98765/consumption/?order_by=period&period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z&page_size=47",
        text=open("tests/fixtures/elec-one-day-usage.json", "r").read(),
    )


@pytest.fixture(name="mock_gas_rates")
def fixture_mock_gas_rates(requests_mock):
    """
    Mock response for gas standard unit rates & standing charge
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/gas-tariffs/G-1R-FIX-12M-20-09-21-H/standard-unit-rates/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/gas-standard-rate.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/gas-tariffs/G-1R-FIX-12M-20-09-21-H/standing-charges/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/gas-standing-charges.json", "r").read(),
    )


@pytest.fixture(name="mock_elec_consumption_five_days")
def fixture_mock_elec_consumption_five_days(requests_mock):
    """
    Mock responses for five days' electricity usage
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/electricity-meter-points/12345/meters/67890/consumption/?order_by=period&period_from=2021-01-14T00:00:00Z&period_to=2021-01-18T23:59:59Z&page_size=239",
        text=open("tests/fixtures/elec-five-day-usage.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/electricity-tariffs/E-1R-FIX-12M-20-09-21-H/standard-unit-rates/?period_from=2021-01-14T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/elec-standard-rate.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/electricity-tariffs/E-1R-FIX-12M-20-09-21-H/standing-charges/?period_from=2021-01-14T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/elec-standing-charges.json", "r").read(),
    )


# TODO: Make this dynamic
@pytest.fixture(name="mock_elec_consumption")
def fixture_mock_elec_consumption(requests_mock):
    """
    Mock responses for five days' electricity usage
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/electricity-meter-points/12345/meters/67890/consumption/?order_by=period&period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z&page_size=47",
        text=open("tests/fixtures/elec-one-day-usage.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/AGILE-18-02-21/electricity-tariffs/E-1R-AGILE-18-02-21-H/standard-unit-rates/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/agile-cost-one-day.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/electricity-meter-points/12345/meters/67890/consumption/?order_by=period&period_from=2021-01-01T00:00:00Z&period_to=2021-01-18T23:59:59Z&page_size=863",
        text=open("tests/fixtures/elec-one-day-usage.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/AGILE-18-02-21/electricity-tariffs/E-1R-AGILE-18-02-21-H/standard-unit-rates/?period_from=2021-01-01T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/agile-cost-one-day.json", "r").read(),
    )


@pytest.fixture(name="mock_gas_consumption")
def fixture_mock_gas_consumption(requests_mock):
    """
    Mock responses for five days' gas usage
    """
    requests_mock.get(
        "https://api.octopus.energy/v1/gas-meter-points/54321/meters/98765/consumption/?order_by=period&period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z&page_size=47",
        text=open("tests/fixtures/elec-one-day-usage.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/gas-meter-points/54321/meters/98765/consumption/?order_by=period&period_from=2021-01-01T00:00:00Z&period_to=2021-01-18T23:59:59Z&page_size=863",
        text=open("tests/fixtures/elec-one-day-usage.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/gas-tariffs/G-1R-FIX-12M-20-09-21-H/standard-unit-rates/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/gas-standard-rate.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/gas-tariffs/G-1R-FIX-12M-20-09-21-H/standing-charges/?period_from=2021-01-18T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/gas-standing-charges.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/gas-tariffs/G-1R-FIX-12M-20-09-21-H/standard-unit-rates/?period_from=2021-01-01T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/gas-standard-rate.json", "r").read(),
    )
    requests_mock.get(
        "https://api.octopus.energy/v1/products/FIX-12M-20-09-21/gas-tariffs/G-1R-FIX-12M-20-09-21-H/standing-charges/?period_from=2021-01-01T00:00:00Z&period_to=2021-01-18T23:59:59Z",
        text=open("tests/fixtures/gas-standing-charges.json", "r").read(),
    )


apps/octocost/octocost.py
import datetime
import json
import re

import dateutil.parser
import pytz
import requests
from appdaemon.plugins.hass import hassapi as hass

BASEURL = "https://api.octopus.energy/v1"


class OctoCostToo(hass.Hass):
    def initialize(self):
        self.auth = self.args["auth"]
        self.mpan = self.args["mpan"]
        self.serial = self.args["serial"]
        # We probably shouldn't be making API calls when initialising
        # self.region = self.args.get("region", self.find_region(self.mpan))
        self.region = self.args["region"].upper()
        self.elec_start_date = datetime.date.fromisoformat(str(self.args["start_date"]))
        self.comparison_tariff = self.args.get("comparison_tariff", None)
        self.gas = self.args.get("gas", None)
        if self.gas:
            self.gas_tariff = self.gas.get("gas_tariff", None)
            self.mprn = self.gas.get("mprn", None)
            self.gas_serial = self.gas.get("gas_serial", None)
            self.gas_start_date = datetime.date.fromisoformat(
                str(self.gas.get("gas_start_date"))
            )

        self.run_in(
            self.cost_and_usage_callback,
            5,
            use=self.consumption_url(),
            cost=self.tariff_url(),
            date=self.elec_start_date,
        )

        if self.comparison_tariff:
            self.run_in(
                self.cost_and_usage_callback,
                6,
                use=self.consumption_url(),
                cost=self.tariff_url(
                    energy="electricity", tariff=self.comparison_tariff
                ),
                date=self.elec_start_date,
            )

        if self.gas:
            self.run_in(
                self.cost_and_usage_callback,
                65,
                use=self.consumption_url("gas"),
                cost=self.tariff_url(energy="gas", tariff=self.gas_tariff),
                date=self.gas_start_date,
            )

        for hour in [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22]:
            self.run_daily(
                self.cost_and_usage_callback,
                datetime.time(hour, 5, 0),
                use=self.consumption_url(),
                cost=self.tariff_url(),
                date=self.elec_start_date,
            )

            self.run_daily(
                self.cost_and_usage_callback,
                datetime.time(hour, 6, 0),
                use=self.consumption_url(),
                cost=self.tariff_url(
                    energy="electricity", tariff=self.comparison_tariff
                ),
                date=self.elec_start_date,
            )

            if self.gas:
                self.run_daily(
                    self.cost_and_usage_callback,
                    datetime.time(hour, 7, 0),
                    use=self.consumption_url("gas"),
                    cost=self.tariff_url(energy="gas", tariff=self.gas_tariff),
                    date=self.gas_start_date,
                )

    @classmethod
    def find_region(cls, mpan):
        url = BASEURL + "/electricity-meter-points/" + str(mpan)
        meter_details = requests.get(url)
        json_meter_details = json.loads(meter_details.text)
        region = str(json_meter_details["gsp"][-1])
        return region

    def tariff_url(self, **kwargs):
        energy = kwargs.get("energy", "electricity")
        tariff = kwargs.get("tariff", "AGILE-18-02-21")
        units = kwargs.get("units", "standard-unit-rates")
        url = f"{BASEURL}/products/{tariff}/{energy}-tariffs/{energy[0].upper()}-1R-{tariff}-{self.region}/{units}/"
        return url

    def consumption_url(self, energy="electricity"):
        meter_point = self.mprn if energy == "gas" else self.mpan
        serial = self.gas_serial if energy == "gas" else self.serial
        url = f"{BASEURL}/{energy}-meter-points/{meter_point}/meters/{serial}/consumption/"
        return url

    def cost_and_usage_callback(self, kwargs):
        self.use_url = kwargs.get("use")
        self.cost_url = kwargs.get("cost")
        self.start_date = kwargs.get("date")
        today = datetime.date.today()
        self.gas = True if "gas" in self.use_url else False
        self.yesterday = today - datetime.timedelta(days=1)
        start_year = datetime.date(today.year, 1, 1)
        start_month = datetime.date(today.year, today.month, 1)
        start_day = self.yesterday

        if today == start_month:
            if today.month == 1:
                start_month = datetime.date(today.year - 1, 12, 1)
            else:
                start_month = datetime.date(today.year, today.month - 1, 1)
        if today == start_year:
            start_year = datetime.date(today.year - 1, 1, 1)

        if self.start_date > start_month:
            start_month = self.start_date

        if self.start_date > start_year:
            start_year = self.start_date

        energy = "gas" if self.gas else "electricity"
        tariff = re.search(r"products/([^/]+)/", self.cost_url).group(1)
        cost_usage = {}
        for period, start in {
            "daily": start_day,
            "monthly": start_month,
            "yearly": start_year,
        }.items():
            # Get the usage and calculate the cost
            (
                cost_usage[f"{period}_usage"],
                cost_usage[f"{period}_cost"],
            ) = self.calculate_cost_and_usage(start=start)
            # Log results for debugging
            self.log(
                "{} {} {} usage: {}".format(
                    period.capitalize(), tariff, energy, cost_usage[f"{period}_usage"]
                ),
                level="INFO",
            )
            self.log(
                "{} {} {} cost: {} p".format(
                    period.capitalize(), tariff, energy, cost_usage[f"{period}_cost"]
                ),
                level="INFO",
            )
            # Set the states
            if self.gas:
                self.set_state(
                    f"sensor.octopus_{period}_gas_usage",
                    state=round(cost_usage[f"{period}_usage"], 2),
                    attributes={"unit_of_measurement": "m3", "icon": "mdi:fire"},
                )
                self.set_state(
                    f"sensor.octopus_{period}_gas_cost",
                    state=round(cost_usage[f"{period}_cost"] / 100, 2),
                    attributes={"unit_of_measurement": "£", "icon": "mdi:cash"},
                )
            else:
                self.set_state(
                    f"sensor.octopus_{period}_usage",
                    state=round(cost_usage[f"{period}_usage"], 2),
                    attributes={"unit_of_measurement": "kWh", "icon": "mdi:flash"},
                )
                if "AGILE" in tariff:
                    self.set_state(
                        f"sensor.octopus_{period}_cost",
                        state=round(cost_usage[f"{period}_cost"] / 100, 2),
                        attributes={"unit_of_measurement": "£", "icon": "mdi:cash"},
                    )
                else:
                    self.set_state(
                        f"sensor.octopus_comparison_{period}_cost",
                        state=round(cost_usage[f"{period}_cost"] / 100, 2),
                        attributes={"unit_of_measurement": "£", "icon": "mdi:cash"},
                    )

    def calculate_count(self, start):
        numberdays = self.yesterday - start
        numberdays = numberdays.days
        return ((numberdays + 1) * 48) - 1

    def calculate_cost_and_usage(self, start):
        usage = 0
        price = 0
        std_chg = 0
        cost = []
        utc = pytz.timezone("UTC")
        request_error = False
        expected_count = self.calculate_count(start=start)
        self.log(f"period_from: {start.isoformat()}T00:00:00Z", level="DEBUG")
        self.log(f"period_to: {self.yesterday.isoformat()}T23:59:59Z", level="DEBUG")

        consump_resp = requests.get(
            url=self.use_url
            + "?order_by=period&period_from="
            + start.isoformat()
            + "T00:00:00Z&period_to="
            + self.yesterday.isoformat()
            + "T23:59:59Z&page_size="
            + str(expected_count),
            auth=(self.auth, ""),
        )

        cost_resp = requests.get(
            url=self.cost_url
            + "?period_from="
            + start.isoformat()
            + "T00:00:00Z&period_to="
            + self.yesterday.isoformat()
            + "T23:59:59Z"
        )

        if consump_resp.status_code != 200:
            self.log(
                f"Error {consump_resp.status_code} getting consumption data: {consump_resp.text}",
                level="ERROR",
            )
            request_error = True
        if cost_resp.status_code != 200:
            self.log(
                f"Error {cost_resp.status_code} getting cost data: {cost_resp.text}",
                level="ERROR",
            )
            request_error = True

        # Don't attempt any further processing as we don't have the data to process
        if request_error:
            return

        # If cost_url does not contain `AGILE`, assume it's a fixed rate and get the standing charge too.
        if "AGILE" not in self.cost_url:
            if "gas-tariffs" in self.cost_url:
                std_chg_url = self.tariff_url(
                    energy="gas", tariff=self.gas_tariff, units="standing-charges"
                )
            else:
                std_chg_url = self.tariff_url(
                    tariff=self.gas_tariff, units="standing-charges"
                )

            standing_chg_resp = requests.get(
                url=std_chg_url
                + "?period_from="
                + start.isoformat()
                + "T00:00:00Z&period_to="
                + self.yesterday.isoformat()
                + "T23:59:59Z"
            )
            if standing_chg_resp.status_code != 200:
                self.log(
                    f"Error {standing_chg_resp.status_code} getting standing charge data: {standing_chg_resp.text}",
                    level="ERROR",
                )
            else:
                standing_chg_json = json.loads(standing_chg_resp.text)
                std_chg = standing_chg_json["results"][0]["value_inc_vat"] * (
                    (self.yesterday - start).days + 1
                )

        consump_json = json.loads(consump_resp.text)
        cost_json = json.loads(cost_resp.text)

        results = consump_json["results"]

        while cost_json["next"]:
            cost.extend(cost_json["results"])
            cost_resp = requests.get(url=cost_json["next"])
            cost_json = json.loads(cost_resp.text)

        cost.extend(cost_json["results"])
        cost.reverse()

        for period in results:
            current_index = results.index(period)
            usage = usage + (results[current_index]["consumption"])
            if "AGILE" not in self.cost_url:
                # Only dealing with gas price which doesn't vary at the moment
                if cost_json["count"] == 1:
                    cost = cost_json["results"][0]["value_inc_vat"]
                    kwh = results[current_index]["consumption"]
                    # Convert consumption from m3 to kWh for gas
                    if "gas-tariffs" in self.cost_url:
                        kwh = kwh * 11.0786

                    price = price + (cost * kwh)
                else:
                    self.log("Error: can only process fixed price gas", level="ERROR")
                    price = 0
            else:
                if (results[current_index]["interval_start"]) != (
                    cost[current_index]["valid_from"]
                ):
                    # Daylight Savings?
                    consumption_date = results[current_index]["interval_start"]
                    if consumption_date.endswith("+01:00"):
                        date_time = dateutil.parser.parse(consumption_date)
                        utc_datetime = date_time.astimezone(utc)
                        utc_iso = utc_datetime.isoformat().replace("+00:00", "Z")
                        if utc_iso == (cost[current_index]["valid_from"]):
                            (results[current_index]["interval_start"]) = utc_iso
                        else:
                            self.log(
                                "UTC Unmatched consumption {}".format(
                                    results[current_index]["interval_start"]
                                )
                                + " / cost {}".format(
                                    cost[current_index]["valid_from"]
                                ),
                                level="ERROR",
                            )
                    else:
                        self.log(
                            "Unmatched consumption {}".format(
                                results[current_index]["interval_start"]
                            )
                            + " / cost {}".format(cost[current_index]["valid_from"]),
                            level="ERROR",
                        )
                price = price + (
                    (cost[current_index]["value_inc_vat"])
                    * (results[current_index]["consumption"])
                )

        # We round because floating point arithmatic leads to some crazy looking figures
        return round(usage, 3), round((price + std_chg), 4)
